libraries/api/src/expressions.js

```diff
@@ -137,10 +137,13 @@ const render = (compiledTemplate, params) => {
   if (ctmpl instanceof Array) {
     return ctmpl.map((value, i) => {
       if (i % 2 === 1) {
-        if (typeof params[value] !== 'string' && typeof params[value] !== 'number') {
-          throw new Error(`Scope expression template expected parameter '${value}' to be a string or number`);
+        switch (typeof params[value]) {
+          case 'string': return params[value];
+          case 'number': return params[value].toString();
+          case 'undefined': return '';  // translate undefined to an empty string
+          default:
+            throw new Error(`Scope expression template expected parameter '${value}' to be a string, number, or undefined`);
         }
-        return params[value];
```

Comment:
This change was required because Dustin and I were not able to have an optional parameter be an empty string. `path-to-regexp` doesn't allow this.

tc-lib-api is applying `req.authorize` because it sees that all of the necessary parameters are present for a route like

```
  route: '/namespaces/:namespace?',
  query: paginateResults.query,
  name: 'listNamespaces',
  scopes: 'index:list-namespaces:<namespace>',
```

In the past that wasn't an issue because we didn't require scopes for this endpoint when namespace is not defined.

The error we were  seeing was `Error: Scope expression template expected parameter 'namespace' to be a string or number`.

/cc @imbstack

Reply:
This used to guard against forgetting a param. The empty string substitution might work but feels dangerous to me. Do we have access to the ? at the end of /namespaces/:namespace? here? We could only allow undefined if it is optional. Maybe another choice would be to subsitute a placeholder string or something? /me waves hands a bit

Also happy to be shown why this isn't actually a problem. We could just add the explanation to this as a comment if so.

Reply:
> Do we have access to the ? at the end of /namespaces/:namespace? 

Maybe there's a way but nothing comes to mind. I looked at the path-to-regexpx [README](https://github.com/pillarjs/path-to-regexp/blob/master/Readme.md) and didn't find a way to do this.

Reply:
Good point, @imbstack - this actually loses some safety.

Another option would be to fix it here:
https://github.com/taskcluster/taskcluster/blob/f79e42cd338091b677292855bb0ada9fe5e1ac87/libraries/api/src/middleware/auth.js#L191-L198
https://github.com/taskcluster/taskcluster/blob/f79e42cd338091b677292855bb0ada9fe5e1ac87/libraries/api/src/utils.js#L2-L14

If that could somehow recognize these "optional" parameters and not return them when calculating useUrlParams, that might achieve what's needed here: tc-lib-api would see that `namespace` isn't always avaliable, so it wouldn't automatically call `req.authorize`, and would leave it to the API method, which can apply the `req.params.namespace || ''` logic.

Reply:
Awesome. Something like that plus a decision on the fakeauth thing and I think we're all set to land :)